Extend GIST to mixtures and non-water solvents

fwaibl opened the pull request:
This pull request extends GIST towards solvents other than water, and, to some extent, towards solvent mixtures (especially salt-water mixtures). Furthermore, it parallelizes the GIST entropy calculation using OpenMP.

**Summary of changes:**
* remove the excludeions option. This does not limit the usability since ions can be excluded also by using "strip".
* Refactor dataset creation into a function.
* If the solvent is not water, warn rather than error. 
* Manage two arrays of datasets. One contains the density of each chemical element contained in the main solvent (i.e., g_O and g_H for water, as well as g_XP for 4-point water models). The other contains molecular densities for a list of solvent molecules that is specified by the user (by default nothing, but can e.g. be g_mol_WAT, g_mol_NA, g_mol_CL). This is specified by the user (option "solventmols"), and the first of "solventmols" is the main solvent (for which atomic densities are computed)
* Change how solvent centers and quaternions are computed. The center is now the center of mass by default, but the old behavior can be obtained by specifying nocom. The quaternion is calculated using three atoms, which can be specified by the user (flag "rigidatoms"), and define a "rigid substructure" of the main solvent. If the user does not specify anything, those atoms are chosen using a simple heuristic (the atom with the most bonds, and two of its bond partners), which results in H1-O-H2 for water.
* Energy calculation is done based on all solvents, entropy, neighbor, and order are based on the main solvent.
* Refactor code for normalization of datasets. During the action, only track the total (neighbors, dTSorient, dTStrans, etc.), then compute the _dens and _norm values in the end. Also, use the polymorphic Dataset3D methods rather than casting to the double and float versions. There are some functions implementing DataSet math. (i.e., scale a dataset or divide by an array)
* Update output printing. There is a fancy (maybe too fancy?) little DataFilePrinter class inside the Action_GIST class. This automatically chooses the format string for int / float columns. Output version is updated to v4. The output is consistent to the old output when not using the new features, but I thought since the columns are now created dynamically, it should be a new output version as well.
* Then entropy calculation is now parallel using OpenMP. All the entropy calculation can be done independently for each voxel, but there is a critical section when writing the output into the output arrays. Furthermore, the large voxel_xyz_ array is not allocated (or at least only the first dimension is allocated) when using skipS_.

**Changes in cpptraj tests**
* update GIST output version to v4
* slight changes in many test files. Several output data are not numerically identical to the old version.
* The PME tests are now passing when compiled with GPU support.
* All tests use the nocom option.

**To-Do**
Several things can/should still be updated.
* Add tests for other solvents and salt-water mixtures. I already tested on some old simulations of mine, and I am already using it for "production" calculations, but there are no small test cases yet.
* Add a check for linearity of the solvent molecules.

**Other problems**
* The neighbor and order calculations are probably not working correctly with non-water solvents. Furthermore, they are not parameterized for other solvents, so they would not make much sense even when implemented without any errors. Therefore, I don't really plan to update them, and instead keep a warning message in the code.
* The code does not work with solvents with 1 or 2 atoms. The option with 1 atom would be easy to implement (simply skip the 6D entropy or set all quaternions to a default value), but the 2 atom option would be more tricky (would need a separate implementation). 

@drroe @tkurtzman what do you think about it? I decided to send a pull request even though it is not completely finished, so I can get some feedback already. Do you have suggestions regarding the changed user interface options, or should I revert some of the changes?
Should we add separate Eww output for components of a solvent mixture? This would make the post-processing a lot easier, but I did not implement it yet because I also did not have it for my publication on salt-water mixtures in GIST, so I am used to work with the single energy output. (The problem is that we need a separate reference energy for each component)

The entropy is currently only calculated for the main solvent. This is no big problem because the 1st order entropy is independent for multiple solvents, so it is cheap to run multiple calculations for multiple solvents. For the energy, this is not an option (at least, it requires many GIST runs)

AmberJenkins commented:
The PGI build in Jenkins failed.

AmberJenkins commented:
The PGI build in Jenkins failed.

AmberJenkins commented:
The PGI build in Jenkins failed.

fwaibl commented:
Should we update the tests to use "DoTest ... -a $TEST_TOLERANCE" also for the full output file? Currently, there is a 0.0001 difference (the last printed digit) in one line of Gist4-Eww-dens.dx and in one line of Gist3-output.dat. Otherwise, I can also update the .save files.

drroe commented:
>  Do you have suggestions regarding the changed user interface options, or should I revert some of the changes?

Why remove the `excludeions` keyword? You can certainly remove it prior to `gist` with `strip`, but then that removes the ion for _every_ subsequent action, not just `gist`. Why not leave `excludeions` for maximum flexibility?

> Should we update the tests to use "DoTest ... -a $TEST_TOLERANCE" also for the full output file? 

It depends on whether you think the error is due something innocuous like floating-point round-off error or not. If the difference is due to the fact that the underlying algorithms have changed (i.e., it will always have the new value), then it's best to update the test values (assuming they have been properly validated). If the difference is because at some point the calculation accumulates round-off error (because numbers become denormalized or something; this happens a lot in curve-fitting for example) then it's appropriate to add a tolerance. Let me know if that makes sense.

AmberJenkins commented:
The PGI build in Jenkins failed.

fwaibl commented:
> Let me know if that makes sense. Also let me know if you'd like me to try to make the changes. I think if you've enabled edits by maintainers that I can work on this PR directly.

I am no expert on git, but as far as I remember you were able to work on my previous pull request.
The test is now compiling, but there is again a 0.00001 difference (probably a summation error - when using OpenMP, the order of summation is not consistent). I will probably just add a test tolerance to the Test_GIST_Order tests as well.

fwaibl commented:
Dear @drroe,
I am having problems re-introducing the `excludeions` option. Would it be possible to just deprecate it for simplicity?
I did some quick tests regarding `excludeions` on the master branch, and it is also not working as expected, so I assume that it was not used much. (GPU crashes, PME produces wrong energies, CPU energy is correct but the Order parameter is different).
If there is a good reason to keep `excludeions` option, I am sure I can get it running within a few days, but otherwise I would prefer to deprecate it.

drroe commented:
> I am having problems re-introducing the excludeions option. Would it be possible to just deprecate it for simplicity?

@fwaibl if it's causing you problems then I think it's fine to get rid of. If it becomes a problem down the road we can see about re-implementing it, but I agree with you that it's probably not worth the effort as it stands. I appreciate you giving it a try though, thanks!

fwaibl commented:
I added some comments and updated the documentation. I also added another regression test for benzene as solvent. Furthermore, I added our papers regarding the extension to multiple solvents and to salt-water mixtures to the GIST citation list. Is this ok with you?
From my side, I think we should be ready to merge this pull request. Is there anything else I should change/fix?

AmberJenkins commented:
The PGI build in Jenkins failed.

drroe commented on src/Action_GIST.cpp:
```diff
@@ -1577,176 +1774,213 @@ void Action_GIST::Print() {
+    infofile_->Printf("Ensemble total solute energy on the grid: %9.5f Kcal/mol \n", SumDataSet(*U_PME_) / NFRAME_);
+  }
+  if (n_linear_solvents_ > 0) {
+    mprintf("GIST warning: %d almost-linear solvent molecules occurred. Maybe choose other \"rigidatoms\".\n");
```
This printf is missing an argument for the `%d` .

fwaibl replied:
Thanks, I meant to print n_linear_solvents.

AmberJenkins commented:
The PGI build in Jenkins failed.

fwaibl commented:
The error in the PGI build seems a bit too much for round-off, since we are not adding a lot of numbers here (it is 10 frames of one molecule). I'll try to get a PGI build locally to debug it.

fwaibl commented:
A more accurate version could be implemented easily by re-normalizing the quaternions at double precision before calculating the angle. However, I would not like to store the quaternions at double precision because the RAM requirement of GIST can sometimes be problematic, and the quaternion array is already one of the main offenders.

drroe commented:
> A more accurate version could be implemented easily by re-normalizing the quaternions at double precision before calculating the angle. However, I would not like to store the quaternions at double precision because the RAM requirement of GIST can sometimes be problematic, and the quaternion array is already one of the main offenders.

I unfortunately don't have ready access to a modern PGI to try to reproduce the problem; this issue doesnt appear with v17 PGI. Is this the code (in Action_GIST::Print()) that leads to the round-off error? :
```C++
double rR = fabs(  voxel_Q_[gr_pt][q1  ] * voxel_Q_[gr_pt][q0  ]
                               + voxel_Q_[gr_pt][q1+1] * voxel_Q_[gr_pt][q0+1]
                               + voxel_Q_[gr_pt][q1+2] * voxel_Q_[gr_pt][q0+2]
                               + voxel_Q_[gr_pt][q1+3] * voxel_Q_[gr_pt][q0+3] ); 
```
If so, can't we just cast all the voxel values to double during the calculation?
```C++
double rR = fabs(  (double)voxel_Q_[gr_pt][q1  ] *(double) voxel_Q_[gr_pt][q0  ]
                               + (double)voxel_Q_[gr_pt][q1+1] * (double)voxel_Q_[gr_pt][q0+1]
```
etc. This can be done anywhere there's potential for round-off error.

fwaibl commented:
> If so, can't we just cast all the voxel values to double during the calculation?
> 
I think this should give us consistent results between compilers. But the angle would not be much more accurate, because the quaternions are not normalized in double precision. A small error in the absolute value of the quaternion leads to a bigger error in the angle because arccos is very steep close to 1.

But, as I said, I expect that this is only relevant very rarely.

drroe commented:
>  But the angle would not be much more accurate, because the quaternions are not normalized in double precision.

I think I see now. It's the values being put into the `voxel_Q_` array itself that are the problem; this could only be solved by making voxel_Q_ double precision, which is a dramatic increase in memory as you have said.

> But, as I said, I expect that this is only relevant very rarely.

If your feeling is that the code is good as-is that's fine. I think we do either (or both) of the following:

1. Increase the test tolerance.
2. Add a different test where the values are not so sensitive.

Doing both might provide the best coverage, but if you just want to do `1` I'm not opposed to that.

fwaibl commented:
I quickly checked, and both compilers give the different result when casting to double. I.e., it seems that GCC is calculating the dot product in float, while PGI is casting to double by default.

fwaibl commented:
Using the following snippet, both compilers print out 0.0152023, which is consistent with the quaternion library, even though the input variables are still in float.
```
#include <iostream>
#include <cmath>

using TYPE = float;

inline double cos_qdist(const double w1, const double x1, const double y1, const double z1, const double w2, const double x2, const double y2, const double z2)
{
    double square_mag1 = w1*w1 + x1*x1 + y1*y1 + z1*z1;
    double square_mag2 = w2*w2 + x2*x2 + y2*y2 + z2*z2;
    double dotprod = w1*w2 + x1*x2 + y1*y2 + z1*z2;
    return dotprod / sqrt(square_mag1*square_mag2);
}

int main()
{
    TYPE w1 = 0.5080507994;
    TYPE x1 = 0.4968648851;
    TYPE y1 = 0.7026906610;
    TYPE z1 = -0.0351498947;

    TYPE w2 = 0.5141307712;
    TYPE x2 = 0.4948045015;
    TYPE y2 = 0.6995792985;
    TYPE z2 = -0.0377739333;

    double norm = w1*w2 + x1*x2 + y1*y2 + z1*z2;
    /* double ang = 2*acos(fabs(w1*w2 + x1*x2 + y1*y2 + z1*z2)); */
    double ang = 2*acos(fabs(cos_qdist(w1, x1, y1, z1, w2, x2, y2, z2)));
    std::cout << norm << " " << ang << std::endl;

    return 0;
}

```

fwaibl commented:
We could also avoid the sqrt by assuming that the quaternions are close to normalized. There is an approximate equation as follows:
```
inline double cos_qdist(const double w1, const double x1, const double y1, const double z1, const double w2, const double x2, const double y2, const double z2)
{
    double square_mag1 = w1*w1 + x1*x1 + y1*y1 + z1*z1;
    double square_mag2 = w2*w2 + x2*x2 + y2*y2 + z2*z2;
    double dotprod = w1*w2 + x1*x2 + y1*y2 + z1*z2;
    double both = square_mag1 * square_mag2;
    double inv_sqrt = 2.0 / (1.0 + both);
    return dotprod * inv_sqrt;
}
```
This still gives the correct result on both compilers. I would have to test how much slowdown this would be compared to the existing implementation.

fwaibl commented:
I am getting a slowdown of 22% (Action_Post) with the more exact quaternion distance. The six-dimensional distance is different by 0.16 kcal/mol over the entire grid. This is surprisingly high (and I will double-check this next week). However, I am not sure whether it is worth the slowdown, because we also have other error sources of similar magnitude (even the output file precision might be in that range).

drroe commented:
> Yeah.  I'd rather be slower and keep the accuracy.  Those errors will be magnified with bigger systems.

Agreed. @fwaibl let me know when you think the code is final and I can merge it.

fwaibl commented:
Now everything seems consistent. If you agree with the new code and there is nothing else to correct, I think we could merge.

tkurtzman commented:
Great!

On Tue, Jun 7, 2022 at 8:21 AM Daniel R. Roe ***@***.***>
wrote:

> Merged #961 <https://github.com/Amber-MD/cpptraj/pull/961> into master.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Amber-MD/cpptraj/pull/961#event-6760790936>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGYKFQ3SOTY4HNMGHYJIA7LVN45DTANCNFSM5SQA4VBQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Sent from My phone
